### PR TITLE
Initial podman build

### PR DESCRIPTION
**Pull request by TomSweeneyRedHat**
Signed-off-by: TomSweeneyRedHat <tsweeney@redhat.com>

Adds the build command to podman which calls 'Buildah bud' under the covers to build a container.
This PR is replacing #30.  I'd a really ugly git mess up and this was the cleanest out.  I hopefully have merged everything appropriately in this PR and have addressed review comments from the previous rounds.

### CODE REVIEW

File: cmd/podman/build.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+			Name:  "signature-policy",
+			Usage: "`pathname` of signature policy file (not usually used)",
+		},
+		cli.StringSliceFlag{
+			Name:  "tag, t",
+			Usage: "`tag` to apply to the built image",
+		},
+		cli.BoolFlag{
+			Name:  "tls-verify",
+			Usage: "Require HTTPS and verify certificates when accessing the registry",
```
Comment (rhatdan) [inline cmd/podman/build.go]: All other usages begin with lower case.

File: cmd/podman/build.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+		},
+	}
+	buildDescription = "This command launches the Buildah command to build an OCI Image. Buildah must be installed for this command to work."
```
Comment (rhatdan) [inline cmd/podman/build.go]: podman build lanches the Buildah command ...

File: cmd/podman/build.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+	}
+)
+
+func buildCmd(c *cli.Context) error {
+
+	budCmdArgs := []string{"bud"}
+
+	if c.IsSet("build-arg") {
+		budCmdArgs = append(budCmdArgs, "--build-arg")
+		budCmdArgs = append(budCmdArgs, c.StringSlice("build-arg")...)
```
Comment (rhatdan) [inline cmd/podman/build.go]: I think this has to be a for loop, no need to check IsSet.
for _, barg := range c.StringSlice("buildah-arg) {
		budCmdArgs = append(budCmdArgs, "--build-arg", barg) 
}

File: cmd/podman/build.go [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+	}
+	if c.IsSet("file") || c.IsSet("f") {
```
Comment (rhatdan) [inline cmd/podman/build.go]: No need to check, for loop will be empty if not set.

File: cmd/podman/build.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+		}
+	}
+	if c.IsSet("format") {
+		budCmdArgs = append(budCmdArgs, "--format", c.String("format"))
+	}
+	if c.IsSet("pull-always") {
+		budCmdArgs = append(budCmdArgs, "--pull-always")
+	}
+	if c.IsSet("quiet") || c.IsSet("q") {
+		boolValue:="False"
```
Comment (rhatdan) [inline cmd/podman/build.go]: All you need is

	if c.IsSet("quiet") {
		budCmdArgs = append(budCmdArgs, "--quiet",  c.Bool("quiet") )
	}

File: cmd/podman/build.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+		if c.Bool("quiet") {
+			boolValue="True"
+		}
+		quietParam:="--quiet=" + boolValue
+		budCmdArgs = append(budCmdArgs, quietParam)
+	}
+	if c.IsSet("runtime") {
+		budCmdArgs = append(budCmdArgs, "--runtime", c.String("runtime"))
+	}
+	if c.IsSet("runtime-flag") {
```
Comment (rhatdan) [inline cmd/podman/build.go]: If runtime-flag is a slice, you need to do a 
for i, rflag := range  c.StringSlice("runtime-flag") {
}

File: cmd/podman/build.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+                budCmdArgs = append(budCmdArgs, c.StringSlice("runtime-flag")...)
+	}
+	if c.IsSet("signature-policy") {
+		budCmdArgs = append(budCmdArgs, "--signature-policy", c.String("signature-policy"))
+	}
+	for _, tagName := range c.StringSlice("tag") {
+		budCmdArgs = append(budCmdArgs, "--tag", tagName)
+	}
+	if c.IsSet("tls-verify") {
+		boolValue:="False"
```
Comment (rhatdan) [inline cmd/podman/build.go]: All you need is
 +		budCmdArgs = append(budCmdArgs, "--tls-verify", c.Bool("tls-verify"))

File: cmd/podman/build.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+		budCmdArgs = append(budCmdArgs, "--tls-verify", boolValue)
+	}
+
+	if len(c.Args()) > 0 {
+		budCmdArgs = append(budCmdArgs, c.Args()...)
+	}
+
+	buildah := "buildah"
+
+	_, err := exec.LookPath("buildah")
```
Comment (rhatdan) [inline cmd/podman/build.go]: Either use the buildah variable from above or use the "buildah" string, not both.

File: completions/bash/podman [hunk 9]
```diff
@@ -684,6 +684,48 @@ _podman_commit() {
     esac
 }
 
+_podman_build() {
```
Comment (rhatdan) [inline completions/bash/podman]: Missing build command below.

File: cmd/podman/build.go (hunk 1 above)
Comment (TomSweeneyRedHat) [reply cmd/podman/build.go]: TY, fixed.

File: cmd/podman/build.go (hunk 2 above)
Comment (TomSweeneyRedHat) [reply cmd/podman/build.go]: changed/fixed.

File: cmd/podman/build.go (hunk 3 above)
Comment (TomSweeneyRedHat) [reply cmd/podman/build.go]: Hmm, I think you're right.  I'll have to check the test to see if I can add something there as it didn't trip up as is.

File: cmd/podman/build.go (hunk 4 above)
Comment (TomSweeneyRedHat) [reply cmd/podman/build.go]: fixed/removed check.

File: cmd/podman/build.go (hunk 5 above)
Comment (TomSweeneyRedHat) [reply cmd/podman/build.go]: I coulda sworn I had double checked the quiet/q and needed both, but just tried a third time and it's working without it.  Removing.

File: cmd/podman/build.go (hunk 6 above)
Comment (TomSweeneyRedHat) [reply cmd/podman/build.go]: fixed

File: completions/bash/podman (hunk 9 above)
Comment (TomSweeneyRedHat) [reply completions/bash/podman]: fixed

File: cmd/podman/build.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+	if c.IsSet("pull-always") {
+		budCmdArgs = append(budCmdArgs, "--pull-always")
+	}
+	if c.IsSet("quiet") {
+		boolValue := "False"
+		if c.Bool("quiet") {
+			boolValue = "True"
+		}
+		quietParam := "--quiet=" + boolValue
+		budCmdArgs = append(budCmdArgs, quietParam)
```
Comment (rhatdan) [inline cmd/podman/build.go]: This one can be greatly simplified to just

budCmdArgs = append(budCmdArgs, "--quiet", c.TBool("Quiet"))

File: cmd/podman/build.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+		budCmdArgs = append(budCmdArgs, c.Args()...)
+	}
+
+	buildah := "buildah"
+
+	_, err := exec.LookPath(buildah)
+	if err != nil {
+		return errors.Wrapf(err, "buildah not found in PATH")
+	}
+	_, err = exec.Command(buildah).Output()
```
Comment (rhatdan) [inline cmd/podman/build.go]: nit Do as single line

if _, err := exec.Command(buildah).Output(); err != nil {
}

Does Output also include Stderr?

**Comment by rhatdan** [comment]
Also update the README.md and the transfer.md

File: cmd/podman/build.go (hunk 10 above)
Comment (TomSweeneyRedHat) [reply cmd/podman/build.go]: I can only get it down to one line.  the quiet param is a bit squirlley, it needs to be "quiet=false" or "quiet" or "quiet=true".  No spaces around the = which happens if you do the append as you have it.

I got it down to:

```
                quietParam := "--quiet=" + strconv.FormatBool(c.Bool("quiet"))
                budCmdArgs = append(budCmdArgs, quietParam)
```

and  you're right, more readable.  Thanks for the thought.

File: cmd/podman/build.go [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+	}
+
+	if len(c.Args()) > 0 {
+		budCmdArgs = append(budCmdArgs, c.Args()...)
+	}
+
+	buildah := "buildah"
+
+	_, err := exec.LookPath(buildah)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (TomSweeneyRedHat) [reply cmd/podman/build.go]: changed.  Someday that golang style will stick in my noggin.

File: cmd/podman/build.go [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+	}
+	for _, tagArg := range c.StringSlice("tag") {
+		budCmdArgs = append(budCmdArgs, "--tag", tagArg)
+	}
+	if c.IsSet("tls-verify") {
+		boolValue := "False"
+		if c.Bool("tls-verify") {
+			boolValue = "True"
+		}
+		budCmdArgs = append(budCmdArgs, "--tls-verify", boolValue)
```
Comment (rhatdan) [inline cmd/podman/build.go]: Should be able to just do
budCmdArgs = append(budCmdArgs, "--tls-verify", c.TBool("tls-verify")

File: cmd/podman/build.go [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+	}
+	if c.IsSet("format") {
+		budCmdArgs = append(budCmdArgs, "--format", c.String("format"))
+	}
+	if c.IsSet("pull-always") {
+		budCmdArgs = append(budCmdArgs, "--pull-always")
+	}
+	if c.IsSet("quiet") {
+		quietParam := "--quiet=" + strconv.FormatBool(c.Bool("quiet"))
+		budCmdArgs = append(budCmdArgs, quietParam)
```
Comment (rhatdan) [inline cmd/podman/build.go]: Ok this looks like a bug in the urfave/cli, which we probably have below also on the other booleans.

File: cmd/podman/build.go (hunk 13 above)
Comment (rhatdan) [reply cmd/podman/build.go]: Actually you need the code like you have for quiet.

Comment (TomSweeneyRedHat) [reply cmd/podman/build.go]: Dope slap, shoulda done that earlier.  Fixed.  Yes I agree, I think the bool handling is a urfave cli  "feature".

**Comment by rhatdan** [comment]
@rh-atomic-bot r+

**Pull request closed without merging**